### PR TITLE
Add Ghost Elements and Conditions to HROM Weights

### PR DESCRIPTION
**Pull request by SADPR**
This PR introduces the ability to include elements and conditions from specified model parts into the Hyper-reduced Order Model (HROM) model part. This functionality provides greater flexibility in HROM applications (for visualization or communication in the case of multiphysics), allowing for the inclusion of 'ghost' elements and conditions.

### Key updates include:

New settings in the rom_settings parameter list: `include_conditions_model_parts_list` and `include_elements_model_parts_list`. These settings enable users to specify the model parts from which conditions and elements should be added to the HROM weights.

Two new utility functions, `AddModelPartConditionsToHROM` and `AddModelPartElementsToHROM`, in RomAuxiliaryUtilities. These functions add the specified conditions and elements from a provided ModelPart to the HROM weights.

### Files Changed:

- `add_custom_utilities_to_python.cpp`: Included the new utility functions for exposure to Python.
- `rom_auxiliary_utilities.cpp` and `rom_auxiliary_utilities.h`: Added the new utility functions.
- `hrom_training_utility.py`: Updated to utilize the new settings and utility functions.

### CODE REVIEW

File: applications/RomApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 1]
```diff
@@ -50,6 +50,8 @@ void AddCustomUtilitiesToPython(pybind11::module& m)
         .def_static("SetHRomComputingModelPart", &RomAuxiliaryUtilities::SetHRomComputingModelPart)
         .def_static("SetHRomVolumetricVisualizationModelPart", &RomAuxiliaryUtilities::SetHRomVolumetricVisualizationModelPart)
         .def_static("GetHRomConditionParentsIds", &RomAuxiliaryUtilities::GetHRomConditionParentsIds)
+        .def_static("AddModelPartConditionsToHROM", &RomAuxiliaryUtilities::AddModelPartConditionsToHROM)
+        .def_static("AddModelPartElementsToHROM", &RomAuxiliaryUtilities::AddModelPartElementsToHROM)
```
Comment (Rbravo555) [inline applications/RomApplication/custom_python/add_custom_utilities_to_python.cpp]: What are these functions doing exactly?

- If, as their name suggests, they ADD either elements or conditions to a list, then why shall we expand manually the list after applying the function? 
- If these functions were called GetIds, the list should not be modified inside the functions. And then, modifying the list after appying the function based on the ids returned, would be the expected behaviour.

File: applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp [hunk 2]
```diff
@@ -335,6 +335,50 @@ std::vector<IndexType> RomAuxiliaryUtilities::GetHRomConditionParentsIds(
...
+    std::vector<IndexType> new_element_ids;
+    auto& r_elem_weights = rHRomWeights["Elements"];
+
+    for (const auto& r_elem : rModelPartWithElementsToInclude.Elements()) {
+        IndexType element_id = r_elem.Id();
+        
+        // Check if the element is already added
+        if (r_elem_weights.find(element_id - 1) == r_elem_weights.end()) {
+            // If it's not added, we add it to the HROM weights and the new_element_ids list
+            r_elem_weights[element_id - 1] = 0; // Adjust the value 0 based on your needs.
```
Comment (Rbravo555) [inline applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp]: When would it make sense to add manually a different value than zero? An approximated reduced system of equation could be achived only if the rest of the weights were also modified, isnt it?

File: applications/RomApplication/custom_utilities/rom_auxiliary_utilities.h [hunk 3]
```diff
@@ -107,6 +107,34 @@ class KRATOS_API(ROM_APPLICATION) RomAuxiliaryUtilities
         const ModelPart& rModelPart,
         const std::map<std::string, std::map<IndexType, double>>& rHRomWeights);
 
+    /**
```
Comment (Rbravo555) [inline applications/RomApplication/custom_utilities/rom_auxiliary_utilities.h:110]: This comments should match the behaviour of the functions. Either they ADD the entities to the list, or just return the ids

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 4]
```diff
@@ -176,18 +195,57 @@ def _GetResidualsProjectedMatrix(self):
...
+
+                # Call the AddModelPartConditionsToHROM function
+                new_conditions = KratosROM.RomAuxiliaryUtilities.AddModelPartConditionsToHROM(
+                    root_model_part, # The complete model part
+                    conditions_to_include_model_part, # The model part containing the conditions to be included
+                    hrom_weights)
+
+                # Add the new conditions to the conditions dict with a null weight
+                for condition_id in new_conditions:
+                    hrom_weights["Conditions"][condition_id] = 0.0
```
Comment (Rbravo555) [inline applications/RomApplication/python_scripts/hrom_training_utility.py:222]: Isnt this duplicating the entries in the already-modified list?

File: applications/RomApplication/python_scripts/hrom_training_utility.py [hunk 5]
```diff
@@ -176,18 +195,57 @@ def _GetResidualsProjectedMatrix(self):
...
+
+                # Call the AddModelPartElementsToHROM function
+                new_elements = KratosROM.RomAuxiliaryUtilities.AddModelPartElementsToHROM(
+                    root_model_part, # The complete model part
+                    elements_to_include_model_part, # The model part containing the elements to be included
+                    hrom_weights)
+
+                # Add the new elements to the elements dict with a null weight
+                for element_id in new_elements:
+                    hrom_weights["Elements"][element_id] = 0.0
```
Comment (Rbravo555) [inline applications/RomApplication/python_scripts/hrom_training_utility.py:241]: same as previous comment

**Review by Rbravo555 — COMMENTED** [review]
Minor checks to be addressed before merging

**Review by RiccardoRossi — COMMENTED** [review]
At a very superficial look from my side ok. 
i leave the detailed review and approval to @Rbravo555

File: applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp [hunk 6]
```diff
@@ -335,6 +335,48 @@ std::vector<IndexType> RomAuxiliaryUtilities::GetHRomConditionParentsIds(
...
+{
+    std::vector<IndexType> new_element_ids;
+    auto& r_elem_weights = rHRomWeights["Elements"];
+
+    for (const auto& r_elem : rModelPartWithElementsToInclude.Elements()) {
+        IndexType element_id = r_elem.Id();
+        
+        // Check if the element is already added
+        if (r_elem_weights.find(element_id - 1) == r_elem_weights.end()) {
+            // If it's not added, we add it to the HROM weights and the new_element_ids list
```
Comment (Rbravo555) [inline applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp]:
```suggestion

```

File: applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp [hunk 7]
```diff
@@ -335,6 +335,48 @@ std::vector<IndexType> RomAuxiliaryUtilities::GetHRomConditionParentsIds(
...
+{
+    std::vector<IndexType> new_condition_ids;
+    auto& r_cond_weights = rHRomWeights["Conditions"];
+
+    for (const auto& r_cond : rModelPartWithConditionsToInclude.Conditions()) {
+        IndexType condition_id = r_cond.Id();
+        
+        // Check if the condition is already added
+        if (r_cond_weights.find(condition_id - 1) == r_cond_weights.end()) {
+            // If it's not added, we add it to the HROM weights and the new_condition_ids list
```
Comment (Rbravo555) [inline applications/RomApplication/custom_utilities/rom_auxiliary_utilities.cpp]:
```suggestion
  
```

**Review by Rbravo555 — APPROVED** [review]

**Pull request merged**